internal/engine/compiler/impl_amd64.go

```diff
@@ -3632,150 +3660,95 @@ func (c *amd64Compiler) compileCopyImpl(isTable bool, srcTableIndex, dstTableInd
-			destinationOffset.register)
-	}
+	c.assembler.CompileMemoryToRegister(amd64.CMPQ,
+		amd64ReservedRegisterForCallEngine, callEngineModuleContextMemorySliceLenOffset, destinationOffset.register)
 	destinationBoundOKJump := c.assembler.CompileJump(amd64.JCC)
-	c.compileExitFromNativeCode(outOfBoundsErrorStatus)
+	c.compileExitFromNativeCode(nativeCallStatusCodeMemoryOutOfBounds)
 	c.assembler.SetJumpTargetOnNext(destinationBoundOKJump)
 
-	// Otherwise, ready to copy the value from source to destination.
-	//
-	// If the copy size equal zero, we skip the entire instructions below.
+	// skip zero size
 	c.assembler.CompileRegisterToConst(amd64.CMPQ, copySize.register, 0)
 	skipJump := c.assembler.CompileJump(amd64.JEQ)
 
-	// If source offet < destination offset: for (i = size-1; i >= 0; i--) dst[i] = src[i];
+	// If dest < source, we can copy forwards
 	c.assembler.CompileRegisterToRegister(amd64.CMPQ, destinationOffset.register, sourceOffset.register)
 	destLowerThanSourceJump := c.assembler.CompileJump(amd64.JLS)
 
-	var scale int16
-	var memToReg, regToMem asm.Instruction
-	if isTable {
-		// For tables, we move 8 bytes at once.
-		memToReg = amd64.MOVQ
-		regToMem = memToReg
-		scale = 8
-	} else {
-		memToReg = amd64.MOVBQZX
-		regToMem = amd64.MOVB
-		scale = 1
-	}
-
-	// If source offet < destination offset: for (i = size-1; i >= 0; i--) dst[i] = src[i];
-	var endJump asm.Node
-	{
-		// sourceOffset -= size.
-		c.assembler.CompileRegisterToRegister(amd64.SUBQ, copySize.register, sourceOffset.register)
-		// destinationOffset -= size.
-		c.assembler.CompileRegisterToRegister(amd64.SUBQ, copySize.register, destinationOffset.register)
-
-		if isTable {
-			// Each element is of type uintptr; 2^3 = 1 << pointerSizeLog2.
-			c.assembler.CompileConstToRegister(amd64.SHLQ, pointerSizeLog2, sourceOffset.register)
-			c.assembler.CompileConstToRegister(amd64.SHLQ, pointerSizeLog2, destinationOffset.register)
-			// destinationOffset += table buffer's absolute address.
-			c.assembler.CompileMemoryToRegister(amd64.ADDQ, tmp, tableInstanceTableOffset, destinationOffset.register)
-			// sourceOffset += table buffer's absolute address.
-			c.assembler.CompileMemoryToRegister(amd64.MOVQ, amd64ReservedRegisterForCallEngine, callEngineModuleContextTablesElement0AddressOffset, tmp)
-			c.assembler.CompileMemoryToRegister(amd64.MOVQ, tmp, int64(srcTableIndex*8), tmp)
-			c.assembler.CompileMemoryToRegister(amd64.ADDQ, tmp, tableInstanceTableOffset, sourceOffset.register)
+	// if source + size < dest, we can copy forwards
+	c.assembler.CompileRegisterToRegister(amd64.MOVQ, destinationOffset.register, tmp)
+	c.assembler.CompileRegisterToRegister(amd64.SUBQ, copySize.register, tmp)
+	c.assembler.CompileRegisterToRegister(amd64.CMPQ, sourceOffset.register, tmp)
+	sourceBoundLowerThanDestJump := c.assembler.CompileJump(amd64.JLS)
+
+	assembleCopySection := func(backwards bool) {
```

Comment:
As i mentioned before, please do not define the closures unless _really_ necessary